tests/syft/core/node/common/plan/plan_test.py

```diff
@@ -203,7 +209,118 @@ def test_plan_batched_execution() -> None:
+
+
+def test_make_plan() -> None:
+    alice = sy.VirtualMachine(name="alice")
+    alice_client = alice.get_root_client()
+
+    @make_plan
+    def add_plan(inp=th.zeros((3))) -> th.Tensor:  # type: ignore
+        return inp + inp
+
+    input_tensor = th.tensor([1, 2, 3])
+    plan_pointer = add_plan.send(alice_client)
+    res = plan_pointer(inp=input_tensor)
+    assert th.equal(res.get()[0], th.tensor([2, 4, 6]))
+
+
+def test_make_plan2() -> None:
+    alice = sy.VirtualMachine(name="alice")
+    alice_client = alice.get_root_client()
+
+    @make_plan
+    def mul_plan(inp=th.zeros((3)), inp2=th.zeros((3))) -> th.Tensor:  # type: ignore
+        return inp * inp2
+
+    t1, t2 = th.tensor([1, 2, 3]), th.tensor([1, 2, 3])
+    plan_pointer = mul_plan.send(alice_client)
+    res = plan_pointer(inp=t1, inp2=t2)
+    assert th.equal(res.get()[0], th.tensor([1, 4, 9]))
+
+
+def test_make_plan_error_no_kwargs() -> None:
+    def assertRaises(exc, obj, methodname, *args) -> None:  # type: ignore
+        with pytest.raises(exc) as e_info:
+            getattr(obj, methodname)(*args)
+        assert str(e_info) != ""
+
+    def test_define_plan():  # type: ignore
+        @make_plan
+        def add_plan(inp):  # type: ignore
+            return inp + inp
+
+    # we can only define a plan with *kwargs* when using the @make_plan decorator, not with args
+    assertRaises(ValueError, test_define_plan, "__call__")
+
+
+def test_mlp_plan() -> None:
+    class MLP(sy.Module):
+        def __init__(self, torch_ref):  # type: ignore
+            super().__init__(torch_ref=torch_ref)
+            self.l1 = self.torch_ref.nn.Linear(784, 100)
+            self.a1 = self.torch_ref.nn.ReLU()
+            self.l2 = self.torch_ref.nn.Linear(100, 10)
+
+        def forward(self, x):  # type: ignore
+            x_reshaped = x.view(-1, 28 * 28)
+            l1_out = self.a1(self.l1(x_reshaped))
+            l2_out = self.l2(l1_out)
+            return l2_out
+
+    def set_params(model, params):  # type: ignore
+        """happens outside of plan"""
+        for p, p_new in zip(model.parameters(), params):
+            p.data = p_new.data
+
+    def cross_entropy_loss(logits, targets, batch_size):  # type: ignore
+        norm_logits = logits - logits.max()
+        log_probs = norm_logits - norm_logits.exp().sum(dim=1, keepdim=True).log()
+        return -(targets * log_probs).sum() / batch_size
+
+    def sgd_step(model, lr=0.1):  # type: ignore
+        with ROOT_CLIENT.torch.no_grad():
+            for p in model.parameters():
+                p.data = p.data - lr * p.grad
+                # Todo: fix this
+                p.grad = th.zeros_like(p.grad.get())
+
+    local_model = MLP(th)  # type: ignore
+
+    @make_plan
+    def train(  # type: ignore
+        xs=th.rand([64 * 3, 1, 28, 28]),
+        ys=th.randint(0, 10, [64 * 3, 10]),
+        params=List(local_model.parameters()),
+    ):
+
+        model = local_model.send(ROOT_CLIENT)
```

Comment:
local_model should be passed in as an argument, no? Otherwise the plan might never be able to find it when the plan is used in the wild somewhere.

Reply:
Yes that would be better. However, I ran into problems with sending sy.modules, by passing them as arguments, because it relies on the object2proto function, which is not implemented for sy.modules

Reply:
So i think one of my next projects might be to make it serializable. If you have any ideas here let me know

Reply:
We are working on serializable nn.Module and then a refactor of Sy.Module for this.